src/OrleansRuntime/LogConsistency/ProtocolServices.cs

```diff
@@ -0,0 +1,256 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Orleans;
+using Orleans.Core;
+using Orleans.LogConsistency;
+using Orleans.MultiCluster;
+using Orleans.Runtime;
+using Orleans.SystemTargetInterfaces;
+using Orleans.GrainDirectory;
+
+namespace Orleans.Runtime.LogConsistency
+{
+    /// <summary>
+    /// Functionality for use by log view adaptors that run distributed protocols. 
+    /// This class allows access to these services to providers that cannot see runtime-internals.
+    /// It also stores grain-specific information like the grain reference, and caches 
+    /// </summary>
+    internal class ProtocolServices : IProtocolServices
+    {
+
+        public GrainReference GrainReference { get { return grain.GrainReference; } }
+
+        private Logger log;
+
+        public IMultiClusterRegistrationStrategy RegistrationStrategy { get; private set; }
+
+        private Grain grain;   // links to the grain that owns this service object
+
+        internal ProtocolServices(Grain gr, Logger log, IMultiClusterRegistrationStrategy strategy)
+        {
+            this.grain = gr;
+            this.log = log;
+            this.RegistrationStrategy = strategy;
+
+            if (!Silo.CurrentSilo.GlobalConfig.HasMultiClusterNetwork)
+                PseudoMultiClusterConfiguration = new MultiClusterConfiguration(DateTime.UtcNow, new string[] { PseudoReplicaId }.ToList());
+        }
+
+
+        public async Task<IProtocolMessage> SendMessage(IProtocolMessage payload, string clusterId)
+        {
+            var silo = Silo.CurrentSilo;
```

Comment:
I started refactoring, but this quickly caused cascading changes that are likely redundant with what you are doing in the other PR, so I will leave it the way it is for now.